devito/symbolics/manipulation.py

```diff
@@ -319,12 +319,96 @@ def reuse_if_untouched(expr, args, evaluate=False):
+    `c>=b` we can safely drop `a` and `c` from the candidate list.
+
+    >>> from devito import Symbol
+    >>> a = Symbol('a')
+    >>> b = Symbol('b')
+    >>> c = Symbol('c')
+    >>> d = Symbol('d')
+    >>> evalrel(max, [a, b, c, d], [[Le(d, a), Ge(c, b)]])
+    MAX(a, c)
     """
+    sfunc = (Min if func is min else Max)  # Choose sympy's Min/Max
+    rfunc = (rmin if func is min else rmax)  # Choose built-in MIN/MAX
+
+    mapper = {}
+    processed = []
+
+    # Form relationals so that RHS has more arguments than LHS:
+    # i.e. (a + d >= b) ---> (b <= a + d)
+    assumptions = [asm if len(asm.rhs.args) > len(asm.lhs.args) else
+                   asm.reversed for asm in assumptions]
+
+    # Expand relations if possible
+    for asm in assumptions:
+        if asm.is_Gte and asm.rhs.is_Add and asm.lhs.is_positive:
+            # If `c >= a + b, {a, b, c} >= 0` then add 'c>=a, c>=a'
+            if all(i.is_positive for i in asm.rhs.args):
+                processed.extend(Ge(asm.lhs, i) for i in asm.rhs.args)
+            # If `c >= a + b, a>= 0, b<=0` then add 'c>=b, c<=a'
+            elif len(asm.rhs.args) == 2:
+                processed.extend(Lt(asm.lhs, i) if i.is_positive else
+                                 Gt(asm.lhs, i) for i in asm.rhs.args)
+            else:
+                processed.append(asm)
+        else:
+            processed.append(asm)
+
+    # Symmetry/ Antisymmetry
+    for asm in processed:
+        if set(asm.args).issubset(input):
+            a0, a1 = asm.args
+            if (((asm.__class__ in (Ge, Gt)) and func is max) or
+               ((asm.__class__ in (Le, Lt)) and func is min)):
+                mapper[a1] = a0
+            elif (((asm.__class__ in (Le, Lt)) and func is max) or
+                  ((asm.__class__ in (Ge, Gt)) and func is min)):
+                mapper[a0] = a1
+
+    # Drop one of symmetric pairs [(a, b) , (b, a)] == > [(a, b)]
+    temp = mapper.copy()
+    for k, v in mapper.items():
+        if v in temp and temp[v] is k:
+            temp.pop(k)
+
+    # Collapse graph paths
+    mapper = transitive_closure(temp)
+    input = [i.subs(mapper) for i in input]
+
+    # Explore simplification opportunities that may have emerged and generate MIN/MAX
+    # expression
     try:
-        bool(min(a, b))  # Can it be evaluated or simplified?
-        return min(a, b)
+        bool(sfunc(*input))  # Can it be evaluated or simplified?
+        exp = sfunc(*input)
+        try:
+            if exp.is_Function:
+                if len(exp.args) < len(input):  # if practically simplified
+                    input = list(exp.args)
+            else:
+                return exp  # return fully simplified expression
+        except TypeError:
+            return rfunc(*input)
+        # OrderedDict for deterministic code-gen
+        input = list(OrderedDict.fromkeys(input))
+        return rfunc(*input)
     except TypeError:
-        return MIN(a, b)
+        return rfunc(*input)
```

Comment:
looks like you use the same `return` in both the try and the except, so it can factorized outside, at the bottom of the function?